README.md
<div align="center">
  <img width="225" src="https://github.com/piotrmurach/strings/blob/master/assets/strings_logo.png" alt="strings logo" />
</div>

# Strings::Truncation

[![Gem Version](https://badge.fury.io/rb/strings-truncation.svg)][gem]
[![Actions CI](https://github.com/piotrmurach/strings-truncation/workflows/CI/badge.svg?branch=master)][gh_actions_ci]
[![Build status](https://ci.appveyor.com/api/projects/status/s8y94c4tvi8mgrh2?svg=true)][appveyor]
[![Maintainability](https://api.codeclimate.com/v1/badges/f7ecb5bf87696e522ccb/maintainability)][codeclimate]
[![Coverage Status](https://coveralls.io/repos/github/piotrmurach/strings-truncation/badge.svg?branch=master)][coverage]
[![Inline docs](http://inch-ci.org/github/piotrmurach/strings-truncation.svg?branch=master)][inchpages]

[gem]: http://badge.fury.io/rb/strings-truncation
[gh_actions_ci]: https://github.com/piotrmurach/strings-truncation/actions?query=workflow%3ACI
[appveyor]: https://ci.appveyor.com/project/piotrmurach/strings-truncation
[codeclimate]: https://codeclimate.com/github/piotrmurach/strings-truncation/maintainability
[coverage]: https://coveralls.io/github/piotrmurach/strings-truncation?branch=master
[inchpages]: http://inch-ci.org/github/piotrmurach/strings-truncation

> Truncate strings with fullwidth characters and ANSI codes.

## Features

* No monkey-patching String class
* Omit text from the start, middle, end or both ends
* Account for fullwidth characters in encodings such as UTF-8, EUC-JP
* Shorten text without whitespaces between words (Chinese, Japanese, Korean etc)
* Preserve ANSI escape codes

## Contents

* [1. Usage](#1-usage)
* [2. API](#2-api)
  * [2.1 configure](#21-configure)
  * [2.2 truncate](#22-truncate)
* [3. Extending String class](#3-extending-string-class)

## Installation

Add this line to your application's Gemfile:

```ruby
gem "strings-truncation"
```

And then execute:

    $ bundle

Or install it yourself as:

    $ gem install strings-truncation

## 1. Usage

Use `truncate` to shorten string to 30 characters by default:

```ruby
strings = Strings::Truncation.new
strings.truncate("I try all things, I achieve what I can.")
# => "I try all things, I achieve w…"
```

As a convenience, you can call `truncate` method directly on a class:

```ruby
Strings::Truncation.truncate("I try all things, I achieve what I can.")
# => "I try all things, I achieve w…"
```

To change the default truncation length, pass an integer as a second argument:

```ruby
strings.truncate("I try all things, I achieve what I can.", 15)
# => "I try all thin…"
```

Or if you want to be more explicit and flexible use `:length` keyword:

```ruby
strings.truncate("I try all things, I achieve what I can.", length: 15)
# => "I try all thin…"
```

You can specify custom omission string in place of default `…`:

```ruby
strings.truncate("I try all things, I achieve what I can.", omission: "...")
# => "I try all things, I achieve..."
```

If you wish to truncate preserving words use a string or regexp as a separator:

```ruby
strings.truncate("I try all things, I achieve what I can.", separator: /\s/)
# => "I try all things, I achieve…"
```

You can omit text from the `start`, `middle`, `end` or both `ends`:

```ruby
strings.truncate("I try all things, I achieve what I can", position: :middle)
# => "I try all thing…ve what I can."
```

You can truncate text with fullwidth characters (Chinese, Japanese, Korean etc):

```ruby
strings.truncate("おはようございます", 8)
# => "おはよ…"
```

As well as truncate text that contains ANSI escape codes:

```ruby
strings.truncate("\e[34mI try all things, I achieve what I can\e[0m", 18)
# => "\e[34mI try all things,\e[0m…"
```

## 2. API

### 2.1 configure

To change default configuration settings at initialization use keyword arguments.

For example, to omit text from the start and separate on a whitespace character do:

```ruby
strings = Strings::Truncation.new(position: :start, separator: /\s/)
```

After initialization, you can use `configure` to change settings inside a block:

```ruby
strings.configure do |config|
  config.length 25
  config.omission "[...]"
  config.position :start
  config.separator /\s/
end
```

Alternatively, you can also use `configure` with keyword arguments:

```ruby
strings.configure(position: :start, separator: /\s/)
```

### 2.2 truncate

By default a string is truncated from the end to maximum length of `30` display columns.

```ruby
strings.truncate("I try all things, I achieve what I can.")
# => "I try all things, I achieve w…"
```

To change the default truncation length, pass an integer as a second argument:

```ruby
strings.truncate("I try all things, I achieve what I can.", 15)
# => "I try all thin…"
```

Or use `:length` keyword to be more explicit:

```ruby
strings.truncate("I try all things, I achieve what I can.", length: 15)
# => "I try all thin…"
```

The default `…` omission character can be replaced using `:omission`:

```ruby
strings.truncate("I try all things, I achieve what I can.", omission: "...")
# => "I try all things, I achieve..."
```

You can omit text from the `start`, `middle`, `end` or both `ends` by specifying `:position`:

```ruby
strings.truncate("I try all things, I achieve what I can", position: :start)
# => "…things, I achieve what I can."

strings.truncate("I try all things, I achieve what I can", position: :middle)
# => "I try all thing…ve what I can."

strings.truncate("I try all things, I achieve what I can", position: :ends)
# => "… all things, I achieve what …"
```

To truncate based on custom character(s) use `:separator` that accepts a string or regular expression:

```ruby
strings.truncate("I try all things, I achieve what I can.", separator: /\s/)
=> "I try all things, I achieve…"
```

You can combine all settings to achieve desired result:

```ruby
strings.truncate("I try all things, I achieve what I can.", length: 20,
                 omission: "...", position: :ends, separator: /\s/)
# => "...I achieve what..."
```

## 3. Extending String class

Though it is highly discouraged to pollute core Ruby classes, you can add the required methods to String class by using refinements.

To include all the **Strings::Truncation** methods, you can load extensions like so:

```ruby
require "strings/truncation/extensions"

using Strings::Truncation::Extensions
```

And then call `truncate` directly on any string:

```ruby
"I try all things, I achieve what I can.".truncate(20, separator: " ")
# => "I try all things, I…"
```

## Development

After checking out the repo, run `bin/setup` to install dependencies. Then, run `rake spec` to run the tests. You can also run `bin/console` for an interactive prompt that will allow you to experiment.

To install this gem onto your local machine, run `bundle exec rake install`. To release a new version, update the version number in `version.rb`, and then run `bundle exec rake release`, which will create a git tag for the version, push git commits and tags, and push the `.gem` file to [rubygems.org](https://rubygems.org).

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/piotrmurach/strings-truncation. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](http://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).

## Code of Conduct

Everyone interacting in the Strings::Truncation project’s codebases, issue trackers, chat rooms and mailing lists is expected to follow the [code of conduct](https://github.com/piotrmurach/strings-truncation/blob/master/CODE_OF_CONDUCT.md).

## Copyright

Copyright (c) 2019 Piotr Murach. See LICENSE for further details.


CHANGELOG.md
# Change log

## [v0.1.0] - 2021-02-23

* Initial implementation and release

[v0.1.0]: https://github.com/piotrmurach/strings-truncation/compare/efb1d50...v0.1.0


spec/spec_helper.rb
# frozen_string_literal: true

if ENV["COVERAGE"] == "true"
  require "simplecov"
  require "coveralls"

  SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter.new([
    SimpleCov::Formatter::HTMLFormatter,
    Coveralls::SimpleCov::Formatter
  ])

  SimpleCov.start do
    command_name "spec"
    add_filter "spec"
  end
end

require "bundler/setup"
require "strings/truncation"

RSpec.configure do |config|
  # Enable flags like --only-failures and --next-failure
  config.example_status_persistence_file_path = ".rspec_status"

  # Disable RSpec exposing methods globally on `Module` and `main`
  config.disable_monkey_patching!

  config.expect_with :rspec do |c|
    c.syntax = :expect
  end
end


spec/perf/truncate_spec.rb
# frozen_string_literal: true

require "rspec-benchmark"
require "active_support/core_ext/string/filters"

RSpec.describe Strings::Truncation do
  include RSpec::Benchmark::Matchers

  it "truncates text slower than ActiveSupport" do
    text = "It is not down on any map; true places never are."
    expect {
      Strings::Truncation.truncate(text, 20)
    }.to perform_slower_than {
      text.truncate(20)
    }.at_most(135).times
  end

  it "truncates text with separator slower than ActiveSupport" do
    text = "It is not down on any map; true places never are."
    separator = /\s/
    expect {
      Strings::Truncation.truncate(text, 20, separator: separator)
    }.to perform_slower_than {
      text.truncate(20, separator: separator)
    }.at_most(55).times
  end

  it "allocates no more than 95 objects" do
    text = "It is not down on any map; true places never are."
    expect {
      Strings::Truncation.truncate(text, 20)
    }.to perform_allocation(95).objects
  end
end


spec/unit/truncate_spec.rb
# frozen_string_literal: true

RSpec.describe Strings::Truncation, "#truncate" do
  it "doesn't truncate for nil length and returns the original text" do
    text = "It is not down on any map; true places never are."

    expect(Strings::Truncation.truncate(text, nil)).to eq(text)
  end

  it "truncates to length of 0 and returns an empty string" do
    text = "It is not down on any map; true places never are."

    expect(Strings::Truncation.truncate(text, 0)).to eq("")
  end

  it "truncates to length of 1 and returns only the omission character" do
    text = "It is not down on any map; true places never are."

    expect(Strings::Truncation.truncate(text, 1)).to eq("…")
  end

  it "doesn't change text for equal length" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, text.length)

    expect(truncation).to eq(text)
  end

  it "doesn't truncate text when length exceeds content" do
    text = "It is not down on any map; true places never are."

    expect(Strings::Truncation.truncate(text, 100)).to eq(text)
  end

  it "goes over default length of 30 characters" do
    text = "It is not down on any map; true places never are."

    expect(Strings::Truncation.truncate(text)).to eq("#{text[0..28]}…")
  end

  it "truncates whole words when separated used" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, separator: " ")

    expect(truncation).to eq("It is not down on any map;…")
  end

  it "truncates on word boundary" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, 21, separator: /\s/)

    expect(truncation).to eq("It is not down on…")
  end

  it "truncates using :length option" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, length: 21)

    expect(truncation).to eq("It is not down on an…")
  end

  it "calls truncate on an instance" do
    text = "It is not down on any map; true places never are."
    strings = Strings::Truncation.new
    truncation = strings.truncate(text, length: 21, separator: " ")

    expect(truncation).to eq("It is not down on…")
  end

  it "truncates with a custom omission" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, 40, omission: "[...]")

    expect(truncation).to eq("It is not down on any map; true pla[...]")
  end

  it "truncates with a custom omission and separator" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, 40, omission: "[...]",
                                                        separator: " ")

    expect(truncation).to eq("It is not down on any map; true[...]")
  end

  it "truncates from an index" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, length: 21, position: 10)

    expect(truncation).to eq("…down on any map; tr…")
  end

  it "truncates text from the end" do
    text = "It is not down on any map; true places never are."
    truncation = Strings::Truncation.truncate(text, 20, position: :end)

    expect(truncation).to eq("It is not down on a…")
  end

  it "fails to recognize position parameter" do
    expect {
      Strings::Truncation.truncate("hello", 3, position: :unknown)
    }.to raise_error(Strings::Truncation::Error,
                     "unsupported position: :unknown")
  end

  context "from the start" do
    [
      ["aaaaabbbbb", "", 0],
      ["aaaaabbbbb", "…", 1],
      ["aaaaabbbbb", "…b", 2],
      ["aaaaabbbbb", "…bbbb", 5],
      ["aaaaabbbbb", "…bbbbb", 6],
      ["aaaaabbbbb", "aaaaabbbbb", 10],
      ["aaaaabbbbb", "...bb", 5, { omission: "..." }],
      ["aaaaabbbbb", "...bbb", 6, { omission: "..." }],
      ["aaaaabbbbbc", "…c", 2],
      ["aaaaabbbbbc", "…bbbc", 5],
      ["aaaaabbbbbc", "…bbbbc", 6],
      ["aaaaabbbbbc", "aaaaabbbbbc", 11],
      ["aaaaabbbbbc", "...bc", 5, { omission: "..." }],
      ["aaaaabbbbbc", "...bbc", 6, { omission: "..." }],
      ["aaa bbb ccc", "…", 1, { separator: " " }],
      ["aaa bbb ccc", "…", 2, { separator: " " }],
      ["aaa bbb ccc", "…", 3, { separator: " " }],
      ["aaa bbb ccc", "…ccc", 4, { separator: " " }],
      ["aaa bbb ccc", "…ccc", 5, { separator: " " }],
      ["aaa bbb ccc", "…ccc", 6, { separator: " " }],
      ["aaa bbb ccc", "…ccc", 7, { separator: " " }],
      ["aaa bbb ccc", "…bbb ccc", 8, { separator: " " }],
      ["aaa bbb ccc", "…bbb ccc", 9, { separator: " " }],
      ["aaa bbb ccc", "…bbb ccc", 10, { separator: " " }],
      ["aaa bbb ccc", "aaa bbb ccc", 11, { separator: " " }]
    ].each do |text, truncated, length, options = {}|
      it "truncates #{text.inspect} at #{length} -> #{truncated.inspect}" do
        strings = Strings::Truncation.new
        options.update(position: :start)
        expect(strings.truncate(text, length, **options)).to eq(truncated)
      end
    end

    it "truncates text at the start" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 20, position: :start)

      expect(truncation).to eq("…e places never are.")
    end

    it "truncates text at the start with separator" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 22, position: :start,
                                                          separator: " ")

      expect(truncation).to eq("…places never are.")
    end
  end

  context "from the end" do
    [
      ["aaaaabbbbb", "", 0],
      ["aaaaabbbbb", "…", 1],
      ["aaaaabbbbb", "a…", 2],
      ["aaaaabbbbb", "aaaa…", 5],
      ["aaaaabbbbb", "aaaaa…", 6],
      ["aaaaabbbbb", "aaaaabbbbb", 10],
      ["aaaaabbbbb", "aa...", 5, { omission: "..." }],
      ["aaaaabbbbb", "aaa...", 6, { omission: "..." }],
      ["aaaaabbbbbc", "a…", 2],
      ["aaaaabbbbbc", "aaaa…", 5],
      ["aaaaabbbbbc", "aaaaa…", 6],
      ["aaaaabbbbbc", "aaaaabbbbbc", 11],
      ["aaaaabbbbbc", "aa...", 5, { omission: "..." }],
      ["aaaaabbbbbc", "aaa...", 6, { omission: "..." }],
      ["aaa bbb ccc", "…", 1, { separator: " " }],
      ["aaa bbb ccc", "…", 2, { separator: " " }],
      ["aaa bbb ccc", "…", 3, { separator: " " }],
      ["aaa bbb ccc", "aaa…", 4, { separator: " " }],
      ["aaa bbb ccc", "aaa…", 5, { separator: " " }],
      ["aaa bbb ccc", "aaa…", 6, { separator: " " }],
      ["aaa bbb ccc", "aaa…", 7, { separator: " " }],
      ["aaa bbb ccc", "aaa bbb…", 8, { separator: " " }]
    ].each do |text, truncated, length, options = {}|
      it "truncates #{text.inspect} at #{length} -> #{truncated.inspect}" do
        strings = Strings::Truncation.new
        options.update(position: :end)
        expect(strings.truncate(text, length, **options)).to eq(truncated)
      end
    end
  end

  context "from both ends" do
    [
      ["aaaaabbbbb", "", 0],
      ["aaaaabbbbb", "…", 1],
      ["aaaaabbbbb", "…", 2],
      ["aaaaabbbbb", "…a…", 3],
      ["aaaaabbbbb", "…ab…", 4],
      ["aaaaabbbbb", "…aab…", 5],
      ["aaaaabbbbb", "…aabb…", 6],
      ["aaaaabbbbb", "aaaaabbbbb", 10],
      ["aaaaabbbbb", "...", 5, { omission: "..." }],
      ["aaaaabbbbb", "...a...", 7, { omission: "..." }],
      ["aaaaabbbbb", "...ab...", 8, { omission: "..." }],
      ["aaaaabbbbbc", "…", 2],
      ["aaaaabbbbbc", "…b…", 3],
      ["aaaaabbbbbc", "…ab…", 4],
      ["aaaaabbbbbc", "…abb…", 5],
      ["aaaaabbbbbc", "…aabb…", 6],
      ["aaaaabbbbbc", "aaaaabbbbbc", 11],
      ["aaaaabbbbbc", "...", 5, { omission: "..." }],
      ["aaaaabbbbbc", "...b...", 7, { omission: "..." }],
      ["aaaaabbbbbc", "...ab...", 8, { omission: "..." }],
      ["aaa bbb ccc", "…", 1, { separator: " " }],
      ["aaa bbb ccc", "…", 2, { separator: " " }],
      ["aaa bbb ccc", "…", 3, { separator: " " }],
      ["aaa bbb ccc", "…", 4, { separator: " " }],
      ["aaa bbb ccc", "…bbb…", 5, { separator: " " }],
      ["aaa bbb ccc", "…bbb…", 6, { separator: " " }],
      ["aaa bbb ccc", "…bbb…", 7, { separator: " " }],
      ["aaa bbb ccc", "…bbb ccc", 8, { separator: " " }],
      ["aaa bbb ccc", "..bbb..", 9, { separator: " ", omission: ".." }],
      ["aaa bbb ccc", "..bbb ccc", 10, { separator: " ", omission: ".." }]
    ].each do |text, truncated, length, options = {}|
      it "truncates #{text.inspect} at #{length} -> #{truncated.inspect}" do
        strings = Strings::Truncation.new
        options.update(position: :ends)
        expect(strings.truncate(text, length, **options)).to eq(truncated)
      end
    end

    it "truncates text from both ends with long omission and separator" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 35, position: :ends,
                                                          omission: "...",
                                                          separator: " ")

      expect(truncation).to eq("...down on any map; true places...")
    end
  end

  context "from the middle" do
    [
      ["aaaaabbbbb", "", 0],
      ["aaaaabbbbb", "…", 1],
      ["aaaaabbbbb", "a…", 2],
      ["aaaaabbbbb", "a…b", 3],
      ["aaaaabbbbb", "aa…b", 4],
      ["aaaaabbbbb", "aa…bb", 5],
      ["aaaaabbbbb", "aaa…bb", 6],
      ["aaaaabbbbb", "aaaaabbbbb", 10],
      ["aaaaabbbbb", "aa...bb", 7, { omission: "..." }],
      ["aaaaabbbbb", "aaa...bb", 8, { omission: "..." }],
      ["aaaaabbbbbc", "a…", 2],
      ["aaaaabbbbbc", "a…c", 3],
      ["aaaaabbbbbc", "aa…c", 4],
      ["aaaaabbbbbc", "aa…bc", 5],
      ["aaaaabbbbbc", "aaa…bc", 6],
      ["aaaaabbbbbc", "aaaaabbbbbc", 11],
      ["aaaaabbbbbc", "aa...bc", 7, { omission: "..." }],
      ["aaaaabbbbbc", "aaa...bc", 8, { omission: "..." }],
      ["aaa bbb ccc", "…", 1, { separator: " " }],
      ["aaa bbb ccc", "…", 2, { separator: " " }],
      ["aaa bbb ccc", "…", 3, { separator: " " }],
      ["aaa bbb ccc", "…", 4, { separator: " " }],
      ["aaa bbb ccc", "…", 5, { separator: " " }],
      ["aaa bbb ccc", "aaa…", 6, { separator: " " }],
      ["aaa bbb ccc", "aaa…ccc", 7, { separator: " " }],
      ["aaa bbb ccc", "aaa…ccc", 8, { separator: " " }]
    ].each do |text, truncated, length, options = {}|
      it "truncates #{text.inspect} at #{length} -> #{truncated.inspect}" do
        strings = Strings::Truncation.new
        options.update(position: :middle)
        expect(strings.truncate(text, length, **options)).to eq(truncated)
      end
    end

    it "truncates text from the middle" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 20, position: :middle)

      expect(truncation).to eq("It is not …ever are.")
    end

    it "truncates text from the middle with a long omission" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 35, position: :middle,
                                                          omission: "[...]")

      expect(truncation).to eq("It is not down [...]aces never are.")
    end

    it "truncates text from the middle with long omission and separator" do
      text = "It is not down on any map; true places never are."
      truncation = Strings::Truncation.truncate(text, 35, position: :middle,
                                                          omission: "[...]",
                                                          separator: " ")

      expect(truncation).to eq("It is not down[...]never are.")
    end
  end
end


lib/strings/truncation/configuration.rb
# frozen_string_literal: true

module Strings
  class Truncation
    # A configuration object used by a Strings::Truncation instance
    #
    # @api private
    class Configuration
      DEFAULT_LENGTH = 30

      DEFAULT_OMISSION = "…"

      DEFAULT_POSITION = 0

      # Create a configuration
      #
      # @api public
      def initialize(length: DEFAULT_LENGTH, omission: DEFAULT_OMISSION,
                     position: DEFAULT_POSITION, separator: nil)
        @length = length
        @omission = omission
        @position = position
        @separator = separator
      end

      # Update current configuration
      #
      # @api public
      def update(length: nil, omission: nil, position: nil, separator: nil)
        @length = length if length
        @omission = omission if omission
        @position = position if position
        @separator = separator if separator
      end

      # The maximum length to truncate to
      #
      # @example
      #   strings = Strings::Truncation.new
      #
      #   strings.configure do |config|
      #     config.length 40
      #   end
      #
      # @param [Integer] number
      #
      # @api public
      def length(number = (not_set = true))
        if not_set
          @length
        else
          @length = number
        end
      end

      # The string to denote omitted content
      #
      # @example
      #   strings = Strings::Truncation.new
      #
      #   strings.configure do |config|
      #     config.omission "..."
      #   end
      #
      # @param [String] string
      #
      # @api public
      def omission(string = (not_set = true))
        if not_set
          @omission
        else
          @omission = string
        end
      end

      # The position of the omission within the string
      #
      # @example
      #   strings = Strings::Truncation.new
      #
      #   strings.configure do |config|
      #     config.position :start
      #   end
      #
      #
      # @param [Symbol] position
      #   the position out of :start, :middle or :end
      #
      # @api public
      def position(position = (not_set = true))
        if not_set
          @position
        else
          @position = position
        end
      end

      # The separator to break the characters into words
      #
      # @example
      #   strings = Strings::Truncation.new
      #
      #   strings.configure do |config|
      #     config.separator /[, ]/
      #   end
      #
      # @param [String|Regexp] separator
      #
      # @api public
      def separator(separator = (not_set = true))
        if not_set
          @separator
        else
          @separator = separator
        end
      end
    end # Configruation
  end # Truncation
end # Strings


lib/strings/truncation.rb
# frozen_string_literal: true

require "forwardable"
require "strscan"
require "strings-ansi"
require "unicode/display_width"

require_relative "truncation/configuration"
require_relative "truncation/version"

module Strings
  class Truncation
    class Error < StandardError; end

    ANSI_REGEXP  = /#{Strings::ANSI::ANSI_MATCHER}/.freeze
    RESET_REGEXP = /#{Regexp.escape(Strings::ANSI::RESET)}/.freeze
    END_REGEXP   = /\A(#{Strings::ANSI::ANSI_MATCHER})*\z/.freeze

    # Global instance
    #
    # @api private
    def self.__instance__
      @__instance__ ||= Truncation.new
    end

    class << self
      extend Forwardable

      delegate %i[truncate] => :__instance__
    end

    # Create a Strings::Truncation instance
    #
    # @example
    #   strings = Strings::Truncation.new(separator: /[,- ]/)
    #
    # @param [Integer] length
    #   the maximum length to truncate to
    # @param [String] omission
    #   the string to denote omitted content
    # @param [String|Integer] position
    #   the position of the omission within the string
    # @param [String|Regexp] separator
    #   the separator to break words on
    #
    # @api public
    def initialize(**options)
      configuration.update(**options)
    end

    # Access configuration
    #
    # @api public
    def configuration
      @configuration ||= Configuration.new
    end

    # Configure truncation
    #
    # @example
    #   strings = Strings::Truncation.new
    #   strings.configure do |config|
    #     config.length 20
    #     config.omission "[...]"
    #     config.position :middle
    #     config.separator /[,- ]/
    #   end
    #
    # @example
    #   strings = Strings::Truncation.new
    #   strings.configure length: 20, omission: "[...]"
    #
    # @yield [Configuration]
    #
    # @api public
    def configure(**options)
      if block_given?
        yield configuration
      else
        configuration.update(**options)
      end
    end

    # Truncate a text at a given length (defualts to 30)
    #
    # @param [String] text
    #   the text to be truncated
    #
    # @param [Integer] truncate_at
    #   the width at which to truncate the text
    #
    # @param [String|Regexp] separator
    #   the character for splitting words
    #
    # @param [String] omission
    #   the string to use for displaying omitted content
    #
    # @param [String|Integer] position
    #   the position in text from which to omit content
    #
    # @example
    #   truncate("It is not down on any map; true places never are.")
    #   # => "It is not down on any map; tr…""
    #
    #   truncate("It is not down on any map; true places never are.", 15)
    #   # => "It is not down…""
    #
    #   truncate("It is not down on any map; true places never are.",
    #            separator: " ")
    #   # => "It is not down on any map;…"
    #
    #   truncate("It is not down on any map; true places never are.", 40,
    #            omission: "[...]")
    #   # => "It is not down on any map; true pla[...]"
    #
    # @api public
    def truncate(text, truncate_at = configuration.length, length: nil,
                 position: configuration.position,
                 separator: configuration.separator,
                 omission: configuration.omission)
      truncate_at = length if length

      return text if truncate_at.nil? || text.bytesize <= truncate_at.to_i

      return "" if truncate_at.zero?

      separator = Regexp.new(separator) if separator

      case position
      when :start
        truncate_start(text, truncate_at, omission, separator)
      when :middle
        truncate_middle(text, truncate_at, omission, separator)
      when :ends
        truncate_ends(text, truncate_at, omission, separator)
      when :end
        truncate_from(0, text, truncate_at, omission, separator)
      when Numeric
        truncate_from(position, text, truncate_at, omission, separator)
      else
        raise Error, "unsupported position: #{position.inspect}"
      end
    end

    private

    # Truncate text at the start
    #
    # @param [String] text
    #   the text to truncate
    # @param [Integer] length
    #   the maximum length to truncate at
    # @param [String] omission
    #   the string to denote omitted content
    # @param [String|Regexp] separator
    #   the pattern or string to separate on
    #
    # @return [String]
    #   the truncated text
    #
    # @api private
    def truncate_start(text, length, omission, separator)
      text_width = display_width(Strings::ANSI.sanitize(text))
      omission_width = display_width(omission)
      return text if text_width == length
      return omission if omission_width == length

      from = [text_width - length, 0].max
      from += omission_width if from > 0
      words, = *slice(text, from, length - omission_width,
                      omission_width: omission_width,
                      separator: separator)

      "#{omission if from > 0}#{words}"
    end

    # Truncate text before the from position and after the length
    #
    # @param [Integer] from
    #   the position to start extracting from
    # @param [String] text
    #   the text to truncate
    # @param [Integer] length
    #   the maximum length to truncate at
    # @param [String] omission
    #   the string to denote omitted content
    # @param [String|Regexp] separator
    #   the pattern or string to separate on
    #
    # @return [String]
    #   the truncated text
    #
    # @api private
    def truncate_from(from, text, length, omission, separator)
      omission_width = display_width(omission)
      length_without_omission = length - omission_width
      length_without_omission -= omission_width if from > 0
      words, stop = *slice(text, from, length_without_omission,
                           omission_width: omission_width,
                           separator: separator)

      "#{omission if from > 0}#{words}#{omission if stop}"
    end

    # Truncate text in the middle
    #
    # @param [String] text
    #   the text to truncate
    # @param [Integer] length
    #   the maximum length to truncate at
    # @param [String] omission
    #   the string to denote omitted content
    # @param [String|Regexp] separator
    #   the pattern or string to separate on
    #
    # @return [String]
    #   the truncated text
    #
    # @api private
    def truncate_middle(text, length, omission, separator)
      text_width = display_width(Strings::ANSI.sanitize(text))
      omission_width = display_width(omission)
      return text if text_width == length
      return omission if omission_width == length

      half_length = length / 2
      rem_length = half_length + length % 2
      half_omission = omission_width / 2
      rem_omission = half_omission + omission_width % 2

      before_words, = *slice(text, 0, half_length - half_omission,
                             omission_width: half_omission,
                             separator: separator)

      after_words, = *slice(text, text_width - rem_length + rem_omission,
                            rem_length - rem_omission,
                            omission_width: rem_omission,
                            separator: separator)

      "#{before_words}#{omission}#{after_words}"
    end

    # Truncate text at both ends
    #
    # @param [String] text
    #   the text to truncate
    # @param [Integer] length
    #   the maximum length to truncate at
    # @param [String] omission
    #   the string to denote omitted content
    # @param [String|Regexp] separator
    #   the pattern or string to separate on
    #
    # @return [String]
    #   the truncated text
    #
    # @api private
    def truncate_ends(text, length, omission, separator)
      text_width = display_width(Strings::ANSI.sanitize(text))
      omission_width = display_width(omission)
      return text if text_width <= length
      return omission if length <= 2 * omission_width

      from = (text_width - length) / 2 + omission_width
      words, stop = *slice(text, from, length - 2 * omission_width,
                           omission_width: omission_width,
                           separator: separator)
      return omission if words.empty?

      "#{omission if from > 0}#{words}#{omission if stop}"
    end

    # Extract number of characters from a text starting at the from position
    #
    # @param [Integer] from
    #   the position to start from
    # @param [Integer] length
    #   the number of characters to extract
    # @param [Integer] omission_width
    #   the width of the omission
    # @param [String|Regexp] separator
    #   the string or pattern to use for splitting words
    #
    # @return [Array<String, Boolean>]
    #   return a substring and a stop flag
    #
    # @api private
    def slice(text, from, length, omission_width: 0, separator: nil)
      scanner = StringScanner.new(text)
      length_with_omission = length + omission_width
      current_length = 0
      start_position = 0
      ansi_reset = false
      visible_char = false
      word_break = false
      stop = false
      words = []
      word = []
      char = nil

      while !(scanner.eos? || stop)
        if scanner.scan(RESET_REGEXP)
          unless scanner.eos?
            words << scanner.matched
            ansi_reset = false
          end
        elsif scanner.scan(ANSI_REGEXP)
          words << scanner.matched
          ansi_reset = true
        else
          if (char =~ separator && start_position <= from) ||
             separator && start_position.zero?
            word_break = start_position != from
          end

          char = scanner.getch
          char_width = display_width(char)
          start_position += char_width
          next if (start_position - char_width) < from

          current_length += char_width

          if char =~ separator
            if word_break
              word_break = false
              current_length = 0
              next
            end
            visible_char = true
            words << word.join
            word.clear
          end

          if current_length <= length || scanner.check(END_REGEXP) &&
                                         current_length <= length_with_omission
            if separator
              word << char unless word_break
            else
              words << char
              visible_char = true
            end
          else
            stop = true
          end
        end
      end

      visible_char = true if !word.empty? && scanner.eos?

      return ["", stop] unless visible_char

      words << word.join if !word.empty? && scanner.eos?

      words << Strings::ANSI::RESET if ansi_reset

      [words.join, stop]
    end

    # Visible width of a string
    #
    # @return [Integer]
    #
    # @api private
    def display_width(string)
      Unicode::DisplayWidth.of(string)
    end
  end # Truncation
end # Strings
